teps/0007-conditions-beta.md

```diff
@@ -0,0 +1,612 @@
+
+```
+apiVersion: tekton.dev/v1alpha1
+kind: Condition
+metadata:
+  name: file-exists
+spec:
+  parameters:
+    - name: "path"
+  resources:
+    - name: workspace
+      type: git
+  check:
+    image: alpine
+    script: 'test -f $(resources.workspace.path)/$(params.path)'
+---
+apiVersion: tekton.dev/v1alpha1
+kind: Task
+metadata:
+  name: file-exists
+spec:
+  parameters:
+    - name: "path"
+  resources:
+    - name: workspace
+      type: git
+  results:
+    - name: skip
+      description: boolean indicating whether guarded Task should be skipped
+  steps:
+  - name: check-file-exists
+    image: alpine
+    script: |
+      if [ -f $(resources.workspace.path)/$(params.path) ]; then
+        echo true | tee /tekton/results/skip
+      else
+        echo false | tee /tekton/results/skip
+      fi
+---
+api: tekton.dev/v1alpha1
+kind: Pipeline
+metadata: 
+    name: generate-file
+    spec: 
+      tasks: 
+        - name: create-file
+          taskRef: 
+            name: create-readme-file 
+        - name: check-file-exists 
+          when:
+            - name: file-exists
+              taskRef: 
+                name: file-exists
+          taskRef:
+            name: echo-file-exists
+          runAfter:
+            - create-file
+```
+Initially, while still supporting `Conditions`, we will add a `Guards` to handle the `Tasks` used as `Guards`. For guarded execution of a given `Task`, users can use either the `Conditions` list or `Guards` list, but not both of them. Over time, we will phase out the `Conditions` and support `Guards` only. 
+
+### Efficiency
+For simple `Guards`, we want to avoid spinning up new `Pods` to check them in order to improve the efficiency of guarded execution of `Tasks`. We can use string expressions for simple `Guard` checking, but we want to avoid adding an opinionated and complex expression language to Tekton API to ensure it can be supported with as many systems as possible. For most use cases, we will use a simple expression syntax from Kubernetes - `MatchExpressions` - which provides `In`, `NotIn` and `Exists` operators. For other cases not covered by MatchExpressions, we will provide a `CelRun` `CustomTask`. 
+
+#### Match Expressions
+We can use expression syntax similar to Kubernetes `MatchExpressions` to evaluate most simple `Guards`. The components of `MatchExpressions` are `Key`, `Operator` and `Values`. 
+
+`Key` is the input for the `Guard` checking - can be from `Parameters` or `Results`. `Values` is an array of string values. `Operator` represents a key's relationship to a set of values. 
+
+Valid [Operators](https://github.com/kubernetes/kubernetes/blob/7f23a743e8c23ac6489340bbb34fa6f1d392db9d/staging/src/k8s.io/apimachinery/pkg/selection/operator.go#L24-L32) are `In`, `NotIn`, `Exists` and `DoesNotExist`. If the `Operator` is `In` or `NotIn`, the `Values` array must be non-empty and if the operator is `Exists` or `DoesNotExist`, the `Values` array must be empty. 
```

Comment:
`matchExpressions` is designed for Kubernetes Labels. Labels consist of pairs of a `Key` and a `Value`. Therefor the first thing is a `key:` in the syntax and the operators `Exists` and `DoesNotExists` means that the `key` exists - independent of what `Value` it has.

Does it make sense for us to adapt the wording and possible the operators? For use, we want to use `Param` or a `TaskResult` as input instead of a `Label-key`. Should we use e.g. `param:` and `taskResult:` or something similar that fits us better? e.g. is more intuitive for end users of Tekton?

I think the `In` and `NotIn` operators fits us good, and users understand them. But if we want to use `Exists` and `DoesNotExists` - we should at least explain what it means in our context - or maybe abandon them - since they are not directly transferrable to our context?

We may also be interested in other operators, e.g. `IsTrue` and `IsNotTrue` / `IsFalse` - and maybe `IsEmpty` and `IsNotEmpty` - but that may also be added later.
These would just be shortcuts for common things that always can be used with `In` in combination of `values: [""]` or `values: ["true"]` / `values: ["false"]`

E.g.

```
name: worskspace-contains-dockerfile
key: $(tasks.dockerfile-check.results.skip)
operator: In
values: ["true"]
```

could be optimized to:

```
name: worskspace-contains-dockerfile
key: $(tasks.dockerfile-check.results.skip)
operator: IsTrue
```

Reply:
TLDR; we might want to start with the operators `In` and `NotIn`

and wait or potentially abandon `Exists` and `DoesNotExists` - in our context.

Reply:
I was thinking we can keep the generic name (maybe rename it to `input`) so that it's flexible to accept different string inputs from different tekton resources using the string reference `'$(x.y.z)'`, so that even as Tekton evolves, we won't have to modify to support them here. For example:

Agreed that we should proceed with `In` and `NotIn`, then we can explore other `Operators` later if needed -- mentioned the examples you provided. 

```yaml
name: check-file-exists
input: '$(tasks.file-exists.results.exists)'
operator: In
values: ['true']
---
name: branch-is-main
input: `$(params.branch)`
operator: In
values: [‘main’]
---
name: always-false
input: ‘false’
operator: In
values: [‘’]
```

Reply:
yes, `input:` sounds better than `key:` in our context.